Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+    if (approved) {
+        auto fd = open(path.characters(), requested_access);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", path, OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptOpenFileResponse ClientConnection::prompt_open_file(String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+    Optional<String> user_picked_file = GUI::FilePicker::get_open_filepath(main_window, "Select file", path_to_view);
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+Messages::FileSystemAccessServer::PromptSaveFileResponse ClientConnection::prompt_save_file(String const& name, String const& ext, String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+    Optional<String> user_picked_file = GUI::FilePicker::get_save_filepath(main_window, name, ext, path_to_view);
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+void ClientConnection::request_download_file(String const& sender, URL const& url, String const& path)
+{
+    // Must receive an absolute path
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    if (m_approved_files.get(path).has_value()) {
+        approved = (m_approved_files.get(path).value() & O_WRONLY) == O_WRONLY;
+    }
+
+    if (!approved) {
+        auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} access to download from \"{}\" to \"{}\"?", sender, url.to_string(), path), "Download Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            m_approved_files.set(path, O_WRONLY);
+        }
```

Comment:
Same as above, maybe use `|=` to unify the granted permissions.

Reply:
Done.